hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieReaderContext.java

```diff
@@ -243,10 +263,7 @@ public boolean castToBoolean(Object value) {
    * @param schema The Avro schema of the record.
    * @return The record key in String.
    */
-  public String getRecordKey(T record, Schema schema) {
-    Object val = getValue(record, schema, RECORD_KEY_METADATA_FIELD);
-    return val.toString();
-  }
+  public abstract String getRecordKey(T record, Schema schema);
```

Comment:
Should a new API be added `getVirtualRecordKey` and keep `getRecordKey` as below so new reader context implementation does not accidentally ignore the record key meta field:
```
public String getRecordKey(T record, Schema schema) {
  if (metaFieldsPopulated) {
    return getValue(record, schema, RECORD_KEY_METADATA_FIELD).toString();
  }
  return getVirtualRecordKey(record, schema);
}
```

Reply:
Good idea, switching to this pattern in the next commit